File: src/waypalm-site/src/main/java/waypalm/site/validation/Captcha.java
package waypalm.site.validation;

import javax.validation.Constraint;
import javax.validation.Payload;
import java.lang.annotation.*;

@Documented
@Constraint(validatedBy = CaptchaValidator.class)
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.RUNTIME)
public @interface Captcha {
    String message() default "{waypalm.site.validation.Captcha.message}";

    Class[] groups() default {};

    Class<? extends Payload>[] payload() default {};

    String challengeField() default "recaptcha_challenge_field";

    String responseField() default "recaptcha_response_field";
}

File: src/waypalm-site/src/main/java/waypalm/site/validation/Captcha.kt
package waypalm.site.validation

import javax.inject.Inject
import waypalm.site.services.RecaptchaValidator
import javax.validation.ConstraintValidator
import javax.validation.ConstraintValidatorContext
import org.springframework.beans.PropertyAccessorFactory

public class CaptchaValidator [Inject](
        val captchaValidator: RecaptchaValidator,
        var challengeField: String = "",
        var responseField: String = ""
)
: ConstraintValidator<Captcha, Any>
{
    public override fun initialize(constraintAnnotation: Captcha?) {
        challengeField = constraintAnnotation!!.challengeField()!!;
        responseField = constraintAnnotation!!.responseField()!!;
    }

    public override fun isValid(value: Any?, context: ConstraintValidatorContext?): Boolean {
        val beanWrapper = PropertyAccessorFactory.forBeanPropertyAccess(value)!!;
        val challenge = beanWrapper.getPropertyValue(challengeField) as String?;
        val response = beanWrapper.getPropertyValue(responseField) as String?;
        return captchaValidator.validate(challenge, response);
    }
}
